### ISSUE TITLE
Ubuntu package

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero-standalone-build#52]
With the switch to a single standalone version, it's possible we should put out an official Ubuntu package. I have no idea what's involved with this, but https://forums.zotero.org/discussion/25317/install-zotero-standalone-from-ubuntu-linux-mint-ppa/p1 and https://github.com/smathot/zotero_installer are possibly relevant.

I'm not going to set this up, but if it can be made simple enough I'm happy to add it to the existing deploy scripts.

**Comment by dstillman** [comment]
Or, alternatively, package the Linux version using Flatpak or AppImage.

**Comment by nedrichards** [comment]
there's currently a flatpak package of this on flathub - although a native one from source code would be better. As per flathub policy, if you'd like to be responsible, as the developer you'd be more than welcome.

**Comment by Eothred** [comment]
I added a shell script that creates the appimage on [a fork](https://github.com/Eothred/zotero/tree/appimage/packaging/AppImage). It essentially depends on 'convert' (for the icon) and wget which are available on most systems I expect. It isn't very properly written (so I will not make a pull request of this in the current state). Someone could probably easily take those steps and write it in a way that it is more properly integrated into the zotero build system.

From the top of my head I know of [canoe](https://github.com/getcanoe/canoe) as another project using npm that builds appimages as releases. (Gruntfile? I know very little about this system in general, sorry) Maybe one can see how it is built there.

**Comment by howthebodyworks** [comment]
@retorquere maintains a repo of [zotero debs](https://github.com/retorquere/zotero-deb).

**Comment by retorquere** [comment]
`.deb` packaging is easy if you have access to the files as laid out just before they'll be tarballed-up., and if you're OK with what is known as a simple repo My scripts are pretty stable now, but the way they package Zotero they're not likely to ever be mainlined into Debian (which is the usual path to get into derivative distros like Ubuntu) because I just put Zotero on disk as it is tarballed, and the zotero-standalone-build fetches a binary (firefox ESR) -- Debian wants everything to be either build from source, or built on an existing debian dependency.

PPA packaging is a mess and is hard to automate. The only benefit is that it's an easier route to get into Ubuntu, but they also want from-source builds, so see former point.

My scripts would be simple to adapt for packaging as part of the build process -- and they'd be stripped pretty far down as they'd not have to detect new versions being available, they wouldn't need to be for both Zotero and Juris-M, etc. I'd be happy to relinquish hosting the `.deb`s. 

The only real thing that would remain is where to host the binaries -- I host them in github releases, but hosting them on S3 would be simpler still. The most important driver for the decision on where to host them is whether you want download stats I think. GH stats are not super great, I don't know about S3. SourceForge is also super easy to set up, and has better download stats (and it's allowed to only use them as a bin hosting service), but SF isn't as popular with the OSS crowd as they once were.

Bintray seemed like the most logical choice but it was finicky to set up, and I hit upload limits during the first two weeks of testing, and I didn't want to be hobbled by the risk that I'd need to put out a release but couldn't.

**Comment by howthebodyworks** [comment]
@retorquere aside: Your various contributions to Zotero, especially [BBT](https://retorque.re/zotero-better-bibtex/) have saved _days_ of my life, thank you.

Also, thanks for the high-speed explanation of debs, Emiliano!  I wonder if we should be avoiding some of the friction that you identified by packaging it the way Ubuntu seems to want pre-built apps distributed these days, i.e. Snap packages. This process I think is not dissimilar to the [flatpak builds](https://github.com/zotero/zotero/issues/1355), which one can also run on Ubuntu.

Snapcraft [explicitly supports pre-built-binaries](https://snapcraft.io/first-snap/pre-built) and seems to include hosting, so that sidesteps the hosting and build questions. I'm not sure if people are have strong commitments to having both formats/pros and cons?

**Comment by retorquere** [comment]
I have no strong opinion on snaps -- I tend to opt for debs when available but that's probably just because I'm used to them. To be clear, debs support pre-build binaries, it's just that PPAs/Debian build rules don't allow them (which is why e.g. Oracle Java is packaged as a shell script that downloads and installs Java during .deb installation).

Snap tools are likely to be more pleasant to work with, and I say this not because I know them but because I can't imagine them being equally frustrating as the .deb tools. It took me ages to find out what was in the end very simple; there's a thousand ways to package debs, and a lot of the docs and tutorials tend to 

1. want to explain everything so even the most complex cases are covered ("If you wish to make an apple pie from scratch, you must first invent the universe"), and
2. be outdated

It is not a pleasant process, but once done, it's set and forget. Because I still want to verify I'm releasing the right thing I'm releasing "manually" but that's really just a commit of an auto-updated config file (and that would be skipped for a real Zotero build, because the check I do is just whether it indeed grabbed the correct version); Travis takes care of the actual build & repo update.

One upside of debs vs snaps is that Chromebook users can install those. Other than that, anything that reliably gets me the latest version of Zotero is fine by me. I have no preference one way or the other.

**Comment by retorquere** [comment]
I did have at one point a setup that used a fork of zotero-standalone-build to build a PPA-compatible package from scratch, but the PPA build infra blocks network access during build, so you can't fetch the firefox binary as part of the build.

Since you can't fetch binaries during build, and you're not allowed to have them as "source", PPAs/Debian mainline will just not work. The same goes for anything that relies on `npm` packages -- they must be existing Debian/Ubuntu packages. I don't see mainlining ever happening for this reason, because it will affect the electron build in the same way.

Self-hosted debs don't have this problem, and neither do snaps/flatpaks.

**Comment by retorquere** [comment]
I can convert (simplify, mainly) my existing script for https://github.com/retorquere/zotero-deb, but I'd need to know

* what is the build environment for releases? Packaging debs requires a debian-based system to create the packages
* what is the stage in the build where the client is laid out on disk as it is about to be tarballed and where is it laid out? Or should I just work from the tarball (not a problem, that's what I do now)
* are you OK with a "simple repo" built with apt-ftparchive?
* where do you intend to host the repo (S3, GH release, something else)?

**Comment by dstillman** [comment]
> what is the build environment for releases?

Not Debian-based, but we could use Docker for that part.

> what is the stage in the build where the client is laid out on disk as it is about to be tarballed and where is it laid out?

Files are available in `dist/` (`$DEST_DIR`) before the [tarball step](https://github.com/zotero/zotero-standalone-build/blob/0b86f6cc1b96132a7d7002d99d605523b373fb8d/build.sh#L542-L547).

> are you OK with a "simple repo" built with apt-ftparchive?

I'm not sure what that means.

> where do you intend to host the repo (S3, GH release, something else)?

The repo in this case is…the thing that gets checked for updates? What form does that take? We tend to keep update manifests on actual webservers in case we need to do conditional updates for any reason (e.g., not serving an update to a particular set of clients). Downloadable files would go on S3 with the rest of the binaries. But I can handle adding the [upload stuff](https://github.com/zotero/zotero-standalone-build/blob/0b86f6cc1b96132a7d7002d99d605523b373fb8d/scripts/build_and_deploy#L117-L160) as long as I know what files need to go where.

Thanks for working on this!

**Comment by retorquere** [comment]
> Files are available in `dist/` (`$DEST_DIR`) before the [tarball step](https://github.com/zotero/zotero-standalone-build/blob/0b86f6cc1b96132a7d7002d99d605523b373fb8d/build.sh#L542-L547).

But they wouldn't be in the Docker env, right? The scripts inside docker could just fetch the tarball as I do now.

> > are you OK with a "simple repo" built with apt-ftparchive?
> 
> I'm not sure what that means.

It means that given a Zotero arch/version, there's only one .deb, not one for each Debian/Ubuntu/whatnot release. In a full repo, you can have one version for Ubuntu 18.10, one for 18.04, etc, but I tried setting that up and it was a major PITA. Given that Zotero only has a single build per arch/version, I don't see any benefit in trying.

> > where do you intend to host the repo (S3, GH release, something else)?
> 
> The repo in this case is…the thing that gets checked for updates? 

Correct, and where the debs are downloaded from

> What form does that take?

Any web server that supports https and that allows files to be downloaded directly (although 301/302 redirects are OK) with a GET; https://hostname/whatever/path/here/can/be/long will work as long as https://hostname/whatever/path/here/can/be/long/InRelease, https://hostname/whatever/path/here/can/be/long/Packages.bz2 etc will work; eg, a simple website hosted in an S3 bucket will work

You can see a full list of assets at https://github.com/retorquere/zotero-deb/releases/tag/apt-get for a simple repo; obviously, a Zotero repo would not host Juris-M binaries. I'd still be maintaining the repo for JM binaries until Frank decides he wants to host them himself, but I'd strip Zotero from the repo.

> We tend to keep update manifests on actual webservers in case we need to do conditional updates for any reason (e.g., not serving an update to a particular set of clients).

A particular set of clients as in OS-dependent? Or another discriminator?

> Downloadable files would go on S3 with the rest of the binaries.

A redirect to S3 would be OK (technically the assets on GH releases are hosted from S3 buckets, and you get sent there by a redirect), but simple repos (and as far as I can tell, full repos) require all URLs to sit on the same base; you can't have just https://hostname1/Packages and https://hostname2/zotero_5.0.73-1_amd64.deb

(the `-1` at the end is a version bump so I can fix things in the packaging while keeping the binaries at the same version).

**Comment by dstillman** [comment]
> But they wouldn't be in the Docker env, right?

No, but we could easily specify the dist directory as a mounted volume. While we could run this as a separate after the normal tarball upload, there's no real reason to redownload and extract the same files.

> Given that Zotero only has a single build per arch/version, I don't see any benefit in trying.

Right.

> A particular set of clients as in OS-dependent? Or another discriminator?

OS version or current Zotero version, mainly. Though I doubt we'd care about distro/version here, and presumably we wouldn't get the Zotero version anyway. So this probably doesn't really matter.

(Basically, this has been useful in the past for things like not serving Zotero above a certain version to users on old versions of macOS, or serving a special manual-upgrade message to clients where the updater was broken.)

**Comment by retorquere** [comment]
Does the build process build both 32 and 64 bit? To finalize the repo you need all packages in place.

**Comment by retorquere** [comment]
It's possible to run them one by one no issue, but

1. You'd need to have a staging area where they come together, and
2. You don't want the repo-refresh after the package build to be done in parallel.

**Comment by retorquere** [comment]
Can I detect in the `dist` directory what architecture (32/64 bit) was being built?

**Comment by retorquere** [comment]
(other than by inspecting the ELF header of `zotero-bin`)

**Comment by dstillman** [comment]
After the `done` [here](https://github.com/zotero/zotero-standalone-build/blob/0b86f6cc1b96132a7d7002d99d605523b373fb8d/build.sh#L548), there'll be Zotero_linux-i686 and Zotero_linux-x86_64 directories.

**Comment by retorquere** [comment]
> OS version or current Zotero version, mainly. Though I doubt we'd care about distro/version here, and presumably we wouldn't get the Zotero version anyway. So this probably doesn't really matter.
> 
> (Basically, this has been useful in the past for things like not serving Zotero above a certain version to users on old versions of macOS

I hate to say this, but this is possible using a full repo. You can have varying versions per target distro. It was a major PITA to get even halfway though, I didn't get it to work last time, and the docs on the process are not set up for easy digestion. Tutorials abound, many of them outdated and conflicting.

> or serving a special manual-upgrade message to clients where the updater was broken.)

Broken updaters are out of scope for debs -- Zotero-updating is disabled in the debs because they're installed as root, updates would come from the dep repo.

> After the `done` [here](https://github.com/zotero/zotero-standalone-build/blob/0b86f6cc1b96132a7d7002d99d605523b373fb8d/build.sh#L548), there'll be `Zotero_linux-i686` and `Zotero_linux-x86_64` directories.

Ah but that script has `$arch` so it could just pass that in.

**Comment by dstillman** [comment]
> I hate to say this, but this is possible using a full repo.

That's OK. There's essentially no chance we'd worry about old distro compatibility the way we've had to worry about macOS and Windows compatibility.

> Broken updaters are out of scope for debs

Right.

If we do decide to split things up and use redirects to S3 for the debs, I'll deal with it. For now we can assume they'll just go to the same place.

**Comment by retorquere** [comment]
> > I hate to say this, but this is possible using a full repo.
> 
> That's OK. There's essentially no chance we'd worry about old distro compatibility the way we've had to worry about macOS and Windows compatibility.

You have no idea what a relief this is 😆 

> If we do decide to split things up and use redirects to S3 for the debs, I'll deal with it. For now we can assume they'll just go to the same place.

That's cool. Then it's mostly done -- at the `done` point there's Zotero_linux-i686 *and* Zotero_linux-x86_64? Not "or"?

**Comment by retorquere** [comment]
Alright, the current script can be started after `done`; when done, it will have `apt/repo` ready to upload. a GPG key needs to be present, I've generated mine using [this](https://github.com/retorquere/zotero-deb#for-packagers----updating-the-packages).

**Comment by retorquere** [comment]
"the current script" being https://github.com/zotero/zotero-standalone-build/pull/73

**Comment by dstillman** [comment]
Resurrecting this in light of https://forums.zotero.org/discussion/comment/394321/#Comment_394321

What's the current feasibility of just integrating the .deb build step into our Linux build script?

**Comment by retorquere** [comment]
As long as either the tarball or the layout pre-tar is available, it can create the package. The main question is whether it is desired to retain older debs. When a new deb is added, the old debs must be present to rebuild the index. I do that now by downloading the release when a new build is needed. 

Another question is version management. De debs follow the zotero release of course, but if a problem is found in the packaging itself, I add an extra sub-minor component to keep the zotero release number while still promoting apt to upgrade. These sub-minors are semiautomatic. I can explain this when we have the infra in place. 

I'm in the last stages of a major simplification of the build script, which should be done in the weekend.

**Comment by retorquere** [comment]
A debian environment is required but that is easily done in a github action (which is where it runs now).

**Comment by dstillman** [comment]
> The main question is whether it is desired to retain older debs. When a new deb is added, the old debs must be present to rebuild the index. I do that now by downloading the release when a new build is needed.

Yes, our packaging scripts do something similar to generate binary diffs. There's a cache folder of recent downloads, and if they're missing it will download them based on an index of recent release versions. But what's the practical upshot in this case? These are the older versions that people can install manually by passing a different version number? And if they're not present those versions become no longer available?

> A debian environment is required but that is easily done in a github action (which is where it runs now).

We'd have to do it in Docker. We do tests via CI but not builds.

**Comment by retorquere** [comment]
> Yes, our packaging scripts do something similar to generate binary diffs. There's a cache folder of recent downloads, and if they're missing it will download them based on an index of recent release versions and a specified number of versions to support. But what's the practical upshot in this case? These are the older versions that people can install manually by passing a different version number to dpkg/apt? And if they're not present those versions become no longer available for manual installs?

Correct. My assumption is that people might sometimes want to downgrade for diagnosis / emergencies. But if not present during the repo build, they'd be unavailable to apt; if the debs are available elsewhere they could be installed using dpkg.

> We'd have to do it in Docker. We do tests via CI but not builds.

Docker would work.

**Comment by dstillman** [comment]
OK, happy to look at whatever you have after you're done working on it. Would be nice to make these part of the official release process.

**Comment by dstillman** [comment]
Do you have any sense of how often most distros are set up to check for and prompt for updates? Should we expect more people to be running out-of-date versions if we make this the recommended installation method?

(One option down the line might be to show an optional notification within Zotero if the current version of Zotero is out of date and the main update system is disabled.)

**Comment by retorquere** [comment]
The default is daily checks.

**Comment by retorquere** [comment]
There's undoubtedly still tweaks to do for the integration, but [this](https://github.com/retorquere/zotero-deb/tree/master/integrated) would be most of what's needed. 

`build.py <path to zotero laid out on disk just before it is tar-bzipped>` will create a temporary directory `build` in which it will prep the deb package, and a directory `repo` that is a ready-to-go repo after it has ran. This can eg be uploaded to an S3 bucket configured for static web hosting, or anything else that makes this directory available under a stable URL.

The build script currently deduces the architecture (32 or 64 bit intel) from the path being passed since I could not find an indicator of the arch inside the directory.

`config.ini` holds configuration for the build process, currently:

```
[maintainer]
email=emiliano.heyns@iris-advies.com
gpgkey=dpkg

[deb]
description=Zotero is a free, easy-to-use tool to help you collect, organize, cite, and share research
dependencies = gnupg
5.0.96.3 = 2
```

The build script will add whatever is in `dependencies` (comma-separated) as dependencies of the package; TBH I don't recall why I added gnupg as a dependency. Previously this also listed `libnss3-dev` which was needed for some systems. The script also automatically add the dependencies for Firefox ESR, except libgcc and lsb-release.

the maintainer email will be added to the package, it can be anything but a support email is suggested. The gpg key for signing the package can be generated using

```
cat << EOF | gpg --gen-key --batch
%no-protection
Key-Type: RSA
Key-Length: 4096
Key-Usage: sign
Name-Real: dpkg
Name-Email: dpkg@iris-advies.com
Expire-Date: 0
%commit
EOF
```

where the `Name-Real` must match the `gpgkey` in `config.ini`. Name-Email can be any valid email address.

The line `5.0.96.3 = 2` is not required, it is an example of patch-handling for the deb-packaging itself. If something in the packaging process is broken and you need to release a new deb package while keeping the Zotero version the same, adding a (higher) number there will cause client systems to download the newly packaged version.

This script can be ran several times to get all combos (32bit, 64bit, 32bit beta, 64bit beta) and will store all those debs in `repo` along with whatever .debs were already present. After each run, the repo is updated to include all .debs that are present there.

I'm sure I've left many questions, fire away.

**Comment by retorquere** [comment]
I'm getting increasing error reports, and github just says it's a problem in apt, not github, which is correct but doesn't help. Whatever comes in its place I'd prefer to be stable from that point on to minimize user disruption. The script is done except for the part where it is injected into the build process, but I don't know where that would be. It would always still be possible to keep the current behavior, which is to probe for new tarballs and download/rebuild when detected, as I do now. That makes the repo process distinct from the zotero build. 

A temporary mitigation could be if I could get an url/domain under zotero.org which could redirect to whatever place I put them for now (probably sourceforge), that way when we do move, it could just redirect to the new place. I could do this with a domain I own, but that would still mean the deb users would have to reconfigure their system.

**Comment by retorquere** [comment]
This is becoming a matter of urgency for me, as I'm getting more and more reports. If none of the options I mentioned above are feasible, I'm going to have to put something together on a domain of my own, but that introduces more disruption for existing users. If I can get at least an URL I could redirect that could later be pointed to the official Zotero hosting, I can do everything needed with no future disruptions. But I need to do something now. Not only are people installing ppa-fixed `apt`, which is very iffy, it doesn't work on all distros, and I think Chromebook users can't use it so are excluded from that fix.

**Comment by dstillman** [comment]
OK, sorry. I haven't reviewed this yet, but for now can we just upload some static files to an S3 bucket? How important is it that it can handle arbitrary/conditional redirects later on?

We'd ultimately make the packages available from download.zotero.org, which is a CDN in front of an S3 bucket. E.g., the tarball is at https://zotero-download.s3.amazonaws.com/client/release/5.0.96.3/Zotero-5.0.96.3_linux-x86_64.tar.bz2. Can we just manually upload files to that same folder?

Are you building debs for the beta?

**Comment by retorquere** [comment]
A simple bucket would do. All a repo of this kind needs is an URL (doesn't need to be top-level) where all repo files can be fetched by appending their filename to the URL. The redirect proposal is just so that we can pick a stable URL, and if at some point the bucket name/physical location would change, the redirect would mean no action is needed from the users. If we already know the stable place, the redirect is unnecessary.

It is possible to upload them to the same bucket you mention. For repo maintenance it might be easier to use a key within the bucket, but it's possible by just using the toplevel of the bucket and mixing them with other downloads. These files could just be placed there, and it'd work, including the `install.sh`. If that's the short-term route, I can just leave the build process as-is for now, and the files could be taken by Zotero from either the GH release or the sourceforge mirror to manually upload to the bucket.

I can also prep things so a GH action would upload the repo to an S3 bucket on change (under a key if need be), test that with a bucket of my own, and we could maybe review that. In that setup, deb releases would trail the tarball releases by no more than two hours (the lowest granularity you can set for scheduled GH actions), but they can also always be ran manually for urgent updates. This is close to my current setup, but since it would be Zotero-only (I presume) and not Juris-M, I would simplify some generalizations (although I can imagine we could offer Frank deb hosting, as I currently do).

I do build beta's, which are updated nightly using a GH action.

**Comment by dstillman** [comment]
Do the files from each release have to be in within the root path? E.g., does `zotero_5.0.96.3_i386.deb` have to be in the same directory or a subdirectory of the parent directory containing `Packages`?

Assuming these files all have to be together, I guess it would look like this:

```
/client/release/apt/Packages
/client/release/apt/5.0.96.3/zotero_5.0.96.3_i386.deb
/client/beta/apt/Packages
/client/beta/apt/5.0.97-beta.57+07df7d0de/5.0.97-beta.57+07df7d0de_i386.deb
```

(If the .deb files could be at arbitrary URLs, we could keep those in the main versioned folders (e.g., `/client/release/5.0.96.3`) and just have the `Packages`/etc. files in `apt` folders under the channel folders.)

Also, if `Packages`/etc. are behind download.zotero.org, we'll have to either set cache control headers on those files to no more than a few minutes so that CloudFront pulls new versions quickly or run a CloudFront invalidation as part of the release process.

I think we'd host install.sh on www.zotero.org alongside our existing download redirector (e.g., `https://www.zotero.org/download/client/dl?channel=release&platform=mac&version=5.0.96.3`), and use something like `https://www.zotero.org/download/client/apt/release/install.sh`.

**Comment by retorquere** [comment]
> Do the files from each release have to be in within the root path? E.g., does `zotero_5.0.96.3_i386.deb` have to be in the same directory or a subdirectory of the parent directory containing `Packages`?

That doesn't seem to be required.

> Assuming these files all have to be together, I guess it would look like this:
> 
> ```
> /client/release/apt/Packages
> /client/release/apt/5.0.96.3/zotero_5.0.96.3_i386.deb
> /client/beta/apt/Packages
> /client/beta/apt/5.0.97-beta.57+07df7d0de/5.0.97-beta.57+07df7d0de_i386.deb
> ```

If that's desired, that can be done, but the beta and the release can co-exist in one repo. Multiple programs can be packaged in a single repo

> (If the .deb files could be at arbitrary URLs, we could keep those in the main versioned folders (e.g., `/client/release/5.0.96.3`) and just have the `Packages`/etc. files in `apt` folders under the channel folders.)

That seems to work. I'll do a test later tonight. 

> I think we'd host install.sh on [www.zotero.org](http://www.zotero.org) alongside our existing download redirector (e.g., `https://www.zotero.org/download/client/dl?channel=release&platform=mac&version=5.0.96.3`), and use something like `https://www.zotero.org/download/client/apt/release/install.sh`.

That should work. The `install.sh` is release-independent.

**Comment by retorquere** [comment]
Do you want to build them during the build of the Zotero client, or from the tarballs?

**Comment by retorquere** [comment]
Another point of note - all debs that you want to have apt-gettable need to be present during repo (re)build. I do that now by using the github action caching, downloading any debs that are missing for whatever reason but which do not need to be rebuilt. If you don't want any of that, detecting a rebuild being needed is not complex (I have that in use, but it's not yet build into the simplified script), and could just download the lot or any limited history, rebuild what needs rebuilding, and publish.

**Comment by dstillman** [comment]
>> Do the files from each release have to be in within the root path? E.g., does zotero_5.0.96.3_i386.deb have to be in the same directory or a subdirectory of the parent directory containing Packages?
>
> That doesn't seem to be required.

Meaning that `Filename` can be an absolute URL, or a relative URL with `..`? Not sure how to square that with https://wiki.debian.org/DebianRepository/Format#Filename

But if we can do this, then yes, I think we'd want to have this layout:

```
/client/release/apt/Packages
/client/release/5.0.96.3/zotero_5.0.96.3_i386.deb
/client/beta/apt/Packages
/client/beta/5.0.97-beta.57%2B07df7d0de/5.0.97-beta.57+07df7d0de_i386.deb
```

There are very occasionally emergencies where we want to pull a release immediately, even if it triggers installation errors, so being able to just rename the folder and invalidate all files for all platforms would be nice.

> Do you want to build them during the build of the Zotero client, or from the tarballs?

There's no real difference. They'd be part of the zotero-standalone-build process, which produces the tarball. But the files will all be present, so no need to re-extract the tarball if that can be avoided.

> Another point of note - all debs that you want to have apt-gettable need to be present during repo (re)build.

That's fine. We can keep all historical debs mirrored on the build machine.

**Comment by retorquere** [comment]
> > > Do the files from each release have to be in within the root path? E.g., does zotero_5.0.96.3_i386.deb have to be in the same directory or a subdirectory of the parent directory containing Packages?
> > 
> > 
> > That doesn't seem to be required.
> 
> Meaning that `Filename` can be an absolute URL, or a relative URL with `..`? Not sure how to square that with https://wiki.debian.org/DebianRepository/Format#Filename

I'm building a simplified repo with `apt-ftparchive`; simplified repo's seem to follow other rules (and the apt source stanza is different for them). The repo building instructions are scattered and seem to range from always-out-of-date tutorials to how-to-design-an-airplane-using-quantum-mechanics details. This is what I managed to get to work consistently. But I'll know whether these relative URLs work when I have the build script updated (doing finishing-up now) and test an actual deployment to S3. Another benefit of these simplified repos is that you don't need to have separate repos for each and every debian/ubuntu/mint/etc release.

> But if we can do this, then yes, I think we'd want to have this layout:
> 
> ```
> /client/release/apt/Packages
> /client/release/5.0.96.3/zotero_5.0.96.3_i386.deb
> /client/beta/apt/Packages
> /client/beta/5.0.97-beta.57%2B07df7d0de/5.0.97-beta.57+07df7d0de_i386.deb
> ```

Can do. So you want history on the beta's too? It's automatic given the current setup, as the build process will just find all debs under `/client`. They do need to be findable from a single directory for `apt-ftparchive` to work.

> There are very occasionally emergencies where we want to pull a release immediately, even if it triggers installation errors, so being able to just rename the folder and invalidate all files for all platforms would be nice.

Got it.

> > Do you want to build them during the build of the Zotero client, or from the tarballs?
> 
> There's no real difference. They'd be part of the zotero-standalone-build process, which produces the tarball. But the files will all be present, so no need to re-extract the tarball if that can be avoided.

I just need to have these available in an environment where the apt tools are available, and the existing debs can be found.

> That's fine. We can keep all historical debs mirrored on the build machine.

All right. So I assume this will all run in a docker container then.

**Comment by retorquere** [comment]
I assume the packaging script does not need to upload the results to S3 (something else will take care of it). The script will be callable either for every separate build, or it can be passed a list of directories with laid out Zotero builds, but it does assume it will not be called concurrently.

**Comment by retorquere** [comment]
Ah and also: I need to know the architecture (`i686` or `x86_64`) to make the package; I did not find this info in the laid out structure, I can take the arch either from the directory name (when unpacking the tarballs, the arch is in the directory name) or by inspecting `zotero-bin`. Preference?

**Comment by dstillman** [comment]
> So you want history on the beta's too?

Not particularly, but it doesn't matter. Possible we'd automatically clean up more than a few recent ones.

> So I assume this will all run in a docker container then.

I suppose. Our Linux build machine isn't Debian-based, so that's probably easiest. We can mount the `staging` folder into the container as a volume.

> I assume the packaging script does not need to upload the results to S3 (something else will take care of it).

Correct.

> The script will be callable either for every separate build, or it can be passed a list of directories with laid out Zotero builds, but it does assume it will not be called concurrently.

I'm assuming we'll run this from build.sh for each architecture after [this line](https://github.com/zotero/zotero-standalone-build/blob/e238d7934d4a24e5d0b4d6ca304b3ce5fde63a96/build.sh#L685).

**Comment by retorquere** [comment]
> I'm assuming we'll run this from build.sh for each architecture after [this line](https://github.com/zotero/zotero-standalone-build/blob/e238d7934d4a24e5d0b4d6ca304b3ce5fde63a96/build.sh#L685).

Can you get me a `find staging -type d` of what the script can be expected to find? I think I have a decent idea, but it will be easier to get the script right.

**Comment by retorquere** [comment]
> > So you want history on the beta's too?
> 
> Not particularly, but it doesn't matter. Possible we'd automatically clean up more than a few recent ones.

Since `apt-ftparchive` just rebuilds exactly to the debs it finds, any cleanup will automatically propagate to the repo.

**Comment by dstillman** [comment]
```
staging/Zotero_linux-x86_64
staging/Zotero_linux-x86_64/fonts
staging/Zotero_linux-x86_64/defaults
staging/Zotero_linux-x86_64/defaults/preferences
staging/Zotero_linux-x86_64/gmp-clearkey
staging/Zotero_linux-x86_64/gmp-clearkey/0.1
staging/Zotero_linux-x86_64/gtk2
staging/Zotero_linux-x86_64/poppler-data
staging/Zotero_linux-x86_64/poppler-data/nameToUnicode
staging/Zotero_linux-x86_64/poppler-data/cidToUnicode
staging/Zotero_linux-x86_64/poppler-data/cMap
staging/Zotero_linux-x86_64/poppler-data/cMap/Adobe-Japan1
staging/Zotero_linux-x86_64/poppler-data/cMap/Adobe-Korea1
staging/Zotero_linux-x86_64/poppler-data/cMap/Adobe-KR
staging/Zotero_linux-x86_64/poppler-data/cMap/Adobe-GB1
staging/Zotero_linux-x86_64/poppler-data/cMap/Adobe-Japan2
staging/Zotero_linux-x86_64/poppler-data/cMap/Adobe-CNS1
staging/Zotero_linux-x86_64/poppler-data/unicodeMap
staging/Zotero_linux-x86_64/extensions
staging/Zotero_linux-x86_64/extensions/zoteroOpenOfficeIntegration@zotero.org
staging/Zotero_linux-x86_64/extensions/zoteroOpenOfficeIntegration@zotero.org/defaults
staging/Zotero_linux-x86_64/extensions/zoteroOpenOfficeIntegration@zotero.org/defaults/preferences
staging/Zotero_linux-x86_64/extensions/zoteroOpenOfficeIntegration@zotero.org/install
staging/Zotero_linux-x86_64/extensions/zoteroOpenOfficeIntegration@zotero.org/chrome
staging/Zotero_linux-x86_64/extensions/zoteroOpenOfficeIntegration@zotero.org/scripts
staging/Zotero_linux-x86_64/extensions/zoteroOpenOfficeIntegration@zotero.org/resource
staging/Zotero_linux-x86_64/extensions/zoteroOpenOfficeIntegration@zotero.org/components
staging/Zotero_linux-x86_64/dictionaries
staging/Zotero_linux-x86_64/icons
staging/Zotero_linux-x86_64/chrome
staging/Zotero_linux-x86_64/chrome/icons
staging/Zotero_linux-x86_64/chrome/icons/default
staging/Zotero_linux-x86_64/components
```

And same for `Zotero_linux-i686`.

**Comment by retorquere** [comment]
And where would `/client` be available to the script?

**Comment by dstillman** [comment]
We can mount that into the container at any path. `/files`?

I assume the script will take a channel string (e.g., `beta`) as an argument?

**Comment by dstillman** [comment]
We can easily pass `arch` (`i686`, `x86_64`), `channel` (`release`, `beta`, `dev`), and `version` (`5.0.96.3`, `5.0.97-beta.37+ddc7be75c`) as necessary.

**Comment by retorquere** [comment]
> But if we can do this, then yes, I think we'd want to have this layout:
> 
> ```
> /client/release/apt/Packages
> /client/release/5.0.96.3/zotero_5.0.96.3_i386.deb
> /client/beta/apt/Packages
> /client/beta/5.0.97-beta.57%2B07df7d0de/5.0.97-beta.57+07df7d0de_i386.deb
> ```

Just so we're clear on this, this gives you two repos, each of which the user would have to run `install.sh` on if they wanted to get both the beta and the binary. It is possible (it is the current setup) to have one repo that hosts both `zotero` and `zotero-beta`. Having them separate means you could have both the beta binary and the release binary be `zotero`, but that means they should not both be installed on the same system. The current implementation installs the beta as `zotero-beta` by renaming `Zotero_linux-x86_64/zotero` in the package to `zotero-beta`, and these can be installed alongside each other.

> We can mount that into the container at any path. `/files`?

That works for me.

> I assume the script will take a channel string (e.g., `beta`) as an argument?

I can add that, but I currently just read that from the `application.ini` that's in staging.

**Comment by retorquere** [comment]
> We can mount that into the container at any path. `/files`?

Can I get a `find /files -type d` for that too?

**Comment by dstillman** [comment]
> The current implementation installs the beta as zotero-beta by renaming Zotero_linux-x86_64/zotero in the package to zotero-beta, and these can be installed alongside each other.

OK, I guess that's fine. We can use a single repo then.

> I can add that, but I currently just read that from the application.ini that's in staging.

Sure.

> Can I get a find /files -type d for that too?

It doesn't exist yet. That would just be the layout I give above, and we'd sync it up to S3. Existing files on S3 would be ignored.

With a single repo, it would be:

```
/client/apt/Packages
/client/release/5.0.96.3/zotero_5.0.96.3_i386.deb
/client/beta/5.0.97-beta.57+07df7d0de/5.0.97-beta.57+07df7d0de_i386.deb
```

So install.sh would point to `https://download.zotero.org/client/apt/`

**Comment by dstillman** [comment]
So then I guess on the [Dev Builds page](https://www.zotero.org/support/dev_builds) we would give an identical install.sh URL and just include an instruction to `apt install zotero-beta` instead of `zotero`?

**Comment by retorquere** [comment]
> So then I guess on the [Dev Builds page](https://www.zotero.org/support/dev_builds) we would give an identical install.sh URL and just include an instruction to `apt install zotero-beta` instead of `zotero`?

Correct, assuming dev builds are the beta's. That's what I do in the BBT test suite.

**Comment by retorquere** [comment]
Took some doing, but relative links work.

**Comment by retorquere** [comment]
OK, so the current script is callable as

```
./build.py --root /files staging
```

and that will leave the deb from that staged directory in the appropriate folder and update `/files/apt`.

The script needs two files as input:

* `mime.xml` which is copied verbatim into the package, and
* `config.ini`, which has information on the maintainer contact, the name of the GPG key to sign the files with, and version bumps.

You can provide the location to these using command line args:

* `--mime`: path to the mime file (doesn't need to be named `mime.xml`), defaults to `mime.xml` in the current directory
* `--config`: path to the config file (doesn't need to be named `config.ini`, defaults to `config.ini` in the current directory
* `--root`: path where the debs and the repo metadata should be written, defaults to the current path

As far as I can tell, that's all. This was tested with py3.9, I can make it compatible with a 2.x version if required, but my preference is 3.x for x >=7. A lot has changed since 2.x, and all for the better. There is currently a dependency on `python-magic`, I can remove that if the arch is passed on the command line, but that would also remove the possibility to build multiple staged directories (but it looks like that is not on the table anyhow, so I can remove that capability).

The docker container needs to have the following ran at start or ran once and cached somehow (I think you can bake the results of these into a docker container, but that's not my AoE):

```
sudo add-apt-repository ppa:mozillateam/ppa -y
sudo apt-get update
sudo apt-get install dpkg-sig fakeroot
```

And someone needs to generate a signing key and make the private part of it available in the docker container.

**Comment by retorquere** [comment]
I have a working setup on a Backblaze B2 bucket, which should be similar to hosting on S3. One problem I did find:

The pointer to the repo must be presented to `apt` as

```
deb <repo url> <distro>
```

These parts must be present, I'm getting around distro-specific packaging (which Zotero does not need, so it's just overhead) by setting the distro to `./`; `apt` will construct an URL from this by unceremoniously pasting them together, so with

```
deb https://downloads.sourceforge.net/project/zotero-deb/ ./
```

you get an URL that looks looks like https://downloads.sourceforge.net/project/zotero-deb/./Packages, which a lot of webservers don't mind understanding as https://downloads.sourceforge.net/project/zotero-deb/Packages, but S3/B2 buckets don't do that.

Sourceforge does, but sourceforge is turning out to be a PITA for hosting as it forces the use of its distributed mirror system, which it updates haphazardly, and `apt` does not like partially updated repos *at all*.

For B2, I've stuck it behind the cloudflare CDN, which can be configured to normalize URLs before picking the assets up from the bucket. But unless we do something similar for the Zotero-provided S3 bucket (maybe the redirector can do this?), having the debs in the versioned directories and the metadata in `/apt` is not going to work, unless there is some option in S3 to normalize URLs. `apt` will not do it.

**Comment by dstillman** [comment]
That appears to work as expected when using CloudFront, which we'd be using:

```
$  curl https://download.zotero.org/tmp/test
OK
$  curl https://download.zotero.org/tmp/./test
OK
```

But we'd be locked into that forever, right? All existing installs would hit that `/./Packages` URL when checking for updates?

How is this related to using the separate versioned directories as opposed to a single folder?

**Comment by dstillman** [comment]
And why use `.` and not just some other string? Sorry, still a bit unclear on how all this fits together.

**Comment by retorquere** [comment]
I've tried that, but with the debs not living under the folder of the Metadata, there's no way around periods ending up in the urls that apt puts together.

**Comment by retorquere** [comment]
> But we'd be locked into that forever, right? All existing installs would hit that `/./Packages` URL when checking for updates?

That is correct. If everything lives under one folder (or the front end makes it look that way) it can be made to look prettier but it'd be essentially the same thing. But this is the case for whatever layout is chosen. Anything that's not from the distro itself, even PPAs, lock in this way. 

> How is this related to using the separate versioned directories as opposed to a single folder?

It's easier to make a single folder look prettier, but that prettyness sits in a file no one ever looks in.

**Comment by retorquere** [comment]
Or is your question whether the layout with the version paths is possible? That is possible. I'll switch the b2 hosting to that format to make sure.

**Comment by dstillman** [comment]
I'm just not clear on the connection between the `.` issue and the versioned paths. I think it would be better to avoid a repo URL without a `.`, so if we can host the debs in subfolders of the repo dir and then use a regular URL, we can just do that. The versioned paths were just to keep all files (apt and otherwise) for a version together, but it's not a requirement.

**Comment by retorquere** [comment]
The simplified repo built using `apt-ftparchive` doesn't seem to tolerate anything but `./`. The file layout works OK, I've just tested that, but it looks like the `./` is some kind of hint to `apt`. If I use

```
deb https://apt.retorque.re/file/zotero-apt/apt ./
```

but anything else gets me either "malformed entry" or "Conflicting distribution"

The docs for apt repo building are absolutely maddening. I have not been able to find a fix for either of these errors, or found how to build a "big boy" repo that would tolerate the regular deb stanza.

Although TBF, I am wondering why the `./` in the URL would be a problem. Nobody has to type that URL anywhere, and the only users who see it is those watch the output scroll by while updating the package list from the command line rather than using the GUI for package management.

**Comment by retorquere** [comment]
> I'm just not clear on the connection between the `./` issue and the versioned paths.

No connection.

**Comment by dstillman** [comment]
> Although TBF, I am wondering why the ./ in the URL would be a problem.

Aesthetic concerns aside, it locks us into a hosting solution that handles the `./`. Which probably isn't an issue — I don't see us stopping using CloudFront anytime soon — but it's a little unsettling.

Are things in a state where I can play with it? I'm not really familiar with any of this, so I think I'll have a better sense of it once I've taken a closer look.

**Comment by retorquere** [comment]
I see. I can look whether I can find out how to create a repo that doesn't need this, but last time it was a few frustrating weeks of googling that got me goose-egg.

Yeah, things are play-ready. Do you mean you want to generate a repo, of have a repo to play with? The latter is at https://apt.retorque.re/file/zotero-apt and within that `apt` (so Packages is at https://apt.retorque.re/file/zotero-apt/apt/Packages) for the script, I'd prefer to do a live walktrough it at all possible.

**Comment by retorquere** [comment]
I've found some recent repo-building tutorials. I'll see what I can do.

**Comment by retorquere** [comment]
OK, I think I can generate a repo that use regular stanzas and can be hosted without needing URL rewriting, but that will only work with this directory layout directly under the location that is named in the stanza:

```
dists
dists/zotero
dists/zotero/main
dists/zotero/main/binary-i386
dists/zotero/main/binary-amd64
pool
pool/main
pool/main/z
pool/main/z/zotero
pool/main/z/zotero-beta
```

which would work with this stanza:

```
deb http://apt.retorque.re/file/zotero-apt zotero main
```

**Comment by retorquere** [comment]
I really need to know whether this is worth investigating. I'm knee-deep in manuals and sifting out deprecated tools, I'm OK with getting this to work (I'm close), but if the layout above is not desirable I need to stop looking into tools and docs to make it happen.

**Comment by dstillman** [comment]
I'm not really sure what that layout is. I don't know how to interpret that without real filenames in real places, like the sample layouts I gave above.

But you don't need to work on this further. I appreciate your continuing to look at it, but if what you have is basically functional in some form, I can just play with it.

**Comment by dstillman** [comment]
(I guess I'm a little confused as to why this is so difficult and undocumented, though. Is hosting a custom apt repo not a normal thing that places do?)

**Comment by retorquere** [comment]
The default non-official-hosting is to use a PPA, but that's not available to Zotero (I tried). Almost all tools to build repos on https://wiki.debian.org/DebianRepository/Setup are deprecated and in various stages of decay. I really wish it were different but it isn't.

this layout:
```
dists
dists/zotero
dists/zotero/main
dists/zotero/main/binary-i386
dists/zotero/main/binary-amd64
pool
pool/main
pool/main/z
pool/main/z/zotero
pool/main/z/zotero-beta
```

isn't actually that important; what's important is that all debs would live inside this structure, not in versioned directories. The two repo directories `dists` and `pool` would be exclusively for the apt repo (they do not need to be in the root), and no apt repo information lives outside of it.

What do you want to play with? A already-built apt repo? That you can find on https://sourceforge.net/projects/zotero-deb/files/. This is a trivial repo (which I previously named a simplified repo, but trivial is the endorsed term for it). I can also make non-trivial repos now, but the most promising tools are:

* aptly, which works, but the author is no longer maintaining it, and I haven't yet found a way to specify the signing key (it will take the default)
* reprepro, which works, but it cannot generate a repo with multiple versions, so it will always only have the latest release and the latest beta. This seems undesirable to me, as downgrades can aid in problem diagnosis or can be used in an emergency when a bug introduced by a new version can't be solved rapidly.

Still looking for ways to get around these limitations. I've skipped tools that have no installable form and would require compiling ourselves, as I think that drags in an unnecessary risk of accidently becoming the maintainer of the tool in question in the process.

**Comment by retorquere** [comment]
I'm also still looking into whether `apt-ftparchive` can be coaxed into generating a trivial repo that does not need `./`.

**Comment by dstillman** [comment]
> what's important is that all debs would live inside this structure, not in versioned directories

Yes, that's fine.

> What do you want to play with? A already-built apt repo?

No, just the code you have in this GitHub repo. If you don't think it's ready and you want to keep working on it, that's fine, but I was just offering to start looking at to see if I can get it to meet our needs if it's in a more-or-less functional state.

**Comment by retorquere** [comment]
I'm doing a test right now on a trivial repo that would not use `./`. It would still be an isolated directory, not versioned directories, but it would not use wrappers around wrappers like `aptly` and `reprepro`, so that should be a fairly stable situation. When that test is successful, I'll strip the code for versioned paths, and then we could look at that.

**Comment by retorquere** [comment]
I have it working; trivial repo, no path rewrites required, a stanza for my bucket would be

```
https://apt.retorque.re/file/zotero-apt repo/
```

meaning it would live in `repo` sub directory in the bucket. It can be any arbitrary path under the bucket. Assuming that (for now at least) it would live under `/client/apt`, I can finish up the script.

**Comment by dstillman** [comment]
> Assuming that (for now at least) it would live under /client/apt, I can finish up the script.

Yes, that's correct. That would be great.

**Comment by retorquere** [comment]
OK, I think it's ready. I'm currently updating 3 repos and they all test green; technically, the repos hosted by me are back up again, but the gh-release hosted repo still needs an unreleased `apt`, which I think isn't a good solution, and while sourceforge works without such patches, but sourceforge updates it mirrors with unpredictable timing and will give errors until the mirrors are synced up. Not great, so I still prefer going with zotero hosting.

The [script](https://github.com/retorquere/zotero-deb/blob/master/build.py) is minimally configurable so that it can still also build jurism packages, which I want to retain until we find a good place for it. It takes as a parameter one or more staged zotero directories to package.

The script expects a [config.ini](https://github.com/retorquere/zotero-deb/blob/master/config.ini); I think it's largely self-explanatory, for the Zotero environment I'd expect wwwroot to be `/files` and repo `/files/client/apt` or `/files/apt` (depending on how `/client` is mounted in the container. `config.ini` needs to be in the directory where the script is started, or its path needs to be passed with `--config`.

[mime.xml](https://github.com/retorquere/zotero-deb/blob/master/mime.xml) needs to be in the directory where the script is started, or its path needs to be passed with `--mime`.

I've added as much comments as I could without cluttering the script, but I'm happy to answer questions. The script needs python 3.8+

**Comment by retorquere** [comment]
New challenge: the sourceforge mirror system does not detect mirrors that are down and will happily redirect you to a dead mirror, so you can end up at a mirror that just times out, which is now happening in my test suite. That leaves the backblaze bucket as my only stable option.

**Comment by retorquere** [comment]
Where do we stand on Zotero hosting of the debs?

**Comment by dstillman** [comment]
Sorry we haven't gotten to this yet. Re: your question in the forums, if it's just a question of cost, we can certainly create an S3 bucket and CloudFront distribution for you, but if we're planning to take these over eventually, is that just an extra manual change that people will need to make? Or is there a smooth migration process, assuming the bucket stays the same?

**Comment by dstillman** [comment]
(Or maybe this is ready-to-go enough that it's something I can just set up in an hour?)

**Comment by retorquere** [comment]
> Sorry we haven't gotten to this yet. Re: your question in the forums, if it's just a question of cost, we can certainly create an S3 bucket (and CloudFront distribution) for you, but if we're planning to take these over eventually, is that just an extra manual change that people will need to make? Or is there a smooth migration process, assuming the bucket stays the same?

That depends under which URL the files are available. If eg a cloudfront redirect/proxy were put in place, it would just mean the cloudfront redirect would be changed, not the URL, and that would be seamless for the user. If the URL changes, action is needed from the user to change along.

> (Or maybe this is ready-to-go enough that it's something I can just set up in an hour?)

that depends on the environment in which the script is ran. I have my current repo refresh broken up in two parts:

1. One part that prepares the environment; arranges that the existing repo content is present, and that any new Zotero versions that must be added are laid out on disk in the same structure that is in the tarball.
2. One part that, given the environment on disk from 1. rebuilds the laid-out client and refreshes the repo

at which point part 1 detects that changes have been made and updates the hosting.

If this is to be ran from the Zotero build, you'd only need part 2, and that is ready-to-run with some configuration only. If the Zotero build environment takes care of part 1, that should be all.

Another way to set this up would be to use [the gh cli](https://cli.github.com/manual/gh_release_download) to just do `gh release download` to the Zotero hosting from the github repo, where the Zotero-hosted apt repo would just be a mirror (but the primary source for users) of the gh release. I could transfer my current repo to Zotero ownership, keep doing the upkeep, and the two infrastructures would be safely separated. I would bring down the sourceforge mirror and rename the GH release, which would alert existing users the hosting needs changes, and after they make the change, it would be Zotero-hosted.

**Comment by retorquere** [comment]
It did turn out that I had just misconfigured b2 + cloudfront, so it should be possible for me to set up a reliable repo that way with minimal cost through the bandwidth alliance.

**Comment by dstillman** [comment]
I'm seeing if I can get the build script to run via Docker.

**Comment by retorquere** [comment]
You will need to have a gpg key in the docker environment to sign the packages (and have it listed in `config.ini`)

**Comment by dstillman** [comment]
Does anything actually check the signature?

Also, what's the deal with this in config.ini?

```
5.0.96.2 = 3
5.0.96.3 = 3
5.0.97.57+07df7d0de = 3
```

**Comment by dstillman** [comment]
Looks like Release.gpg is maybe checked by apt? I assume that's done using the same configured key?

Are the .deb signatures actually checked?

**Comment by retorquere** [comment]
> Does anything actually check the signature?

the `apt` toolchain does, yes. Nothing during build.

> Also, what's the deal with this in config.ini?
> 
> ```
> 5.0.96.2 = 3
> 5.0.96.3 = 3
> 5.0.97.57+07df7d0de = 3
> ```

In the case a problem is found in the packaging process itself after the debs have already been made available, I need to be able to generate a new package-version from the same zotero-version. In the case of `5.0.96.1`, the package version is also `5.0.96.1`, but for `5.0.96.2` there were `5.0.96.2`, `5.0.96.2-1`, `5.0.96.2-2` and finally `5.0.96.2-3`, and apt will have seen these different versions and have upgraded to them. These entries are only present as a sample, I didn't really find 6 issues, but I needed to test the functionality.

> Looks like Release.gpg is maybe checked by apt? I assume that's created here using the same configured key?

correct.

> Are the .deb signatures actually checked?

Yep, by `apt`.

**Comment by retorquere** [comment]
I think I have the [B2 hosting](https://apt.retorque.re/file/zotero-apt/index.html) through the bandwith alliance set up correctly now, so I could in principle switch the hosting to that, and it would be at no cost.

I'm keeping the B2 repo fresh right now, I am not yet announcing it because I don't want to have the users switch too often, but the sourceforge mirror system isn't great -- every few days the automated tests fail because of it without any changes on my end, and that means users are hitting this too. GH releases are stable but require a patched client that certainly ESR users won't have. B2/S3 solves these issues.

BTW, I don't know how much hosting runs for Zotero, but if it's substantial, the [bandwidth alliance](https://www.cloudflare.com/en-gb/bandwidth-alliance/) could be worth looking into.

**Comment by dstillman** [comment]
Meaning you can use CloudFlare's CDN on their free plan and not pay for egress from B2? I couldn't find anything on the CloudFlare pricing page that actually discussed data transfer limits…

I still need to set up GPG but we can probably take this over.

Before we commit to this, how unusual is what we're doing here? How likely is it that something's going to change in Ubuntu that breaks upgrading from existing installs? This issue aside, that's happened at least once before, right?

**Comment by retorquere** [comment]
> Meaning you can use CloudFlare's CDN on their free plan and not pay for egress from B2?

That's my understanding, yes, and since on B2 you don't pay for ingress, the only charged actions are authentication and listing the contents of the bucket, but for me that falls withing the free daily allotment.

> I couldn't find anything on the CloudFlare pricing page that actually discussed data transfer limits…

I've been looking for a while but haven't found any small lettering. The reason I was getting charged earlier was that for each rebuild, I was downloading the repo outside the bandwith alliance path, and while I was developing this I did that frequently.
 
> I still need to set up GPG but we can probably take this over.

Cool.
 
> Before we commit to this, how unusual is what we're doing here? How likely is it that something's going to change in Ubuntu that breaks upgrading from existing installs? 

This repo setup has been in Debian since the beginning as far as I can tell, but outside the Ubuntu's PPA system there hasn't really been a strong culture of outside repo's in Debian, so documentation on it was pretty blurry. What has changed is where it is recommended the public key is stored on the users system. I don't see that changing anymore since each repo now uses its own keychain, the use of the single keychain was the (potential) problem. The current install.sh simply overwrites the keychain, so after they've switched to the zotero-hosted repo, my public key will be gone from their system.

> This issue aside, that's happened at least once before, right? I pretty much only ever use yum-based distros, and custom yum repos are fairly common, but I'm not clear on whether this is as straightforward as those.

I haven't looked into yum repos but was actually thinking about creating RPM packages. But apt repos aren't all that hard in the end. The storage of the key for all (including the official) repos was an issue, but that's no longer in the install.sh

**Comment by retorquere** [comment]
Any news?

**Comment by retorquere** [comment]
Any news on this?

**Comment by retorquere** [comment]
If it's unlikely that this is happening soon(ish), I need to finish setting up the b2-based repo and move everything to that. Sourceforge is frustratingly spotty, and github runs into a debian problem that's not going to be solved quickly; nothing in debian happens quickly. I'd rather not have the users have to move another time when the Zotero repo is live, but sourceforge is unworkable.

**Comment by retorquere** [comment]
What environment does the build run in? Not debian-based, but something else then? Maybe I can get the required tools to work in the actual build environment. I may not need to get familiar with docker.

**Comment by dstillman** [comment]
OK, took some work, but I've got what appears to be a successfully signed repo via Docker.

Need to do some more refinement and testing, but I think this will work. More to come.

**Comment by dstillman** [comment]
Can you explain rebuild.py?

Do I need that, or can I just hard-code `url` and `codename` in install.sh and put it on a server? And then run build.py and an upload script as part of the build process?

**Comment by dstillman** [comment]
Other questions:

1) There's no point to `wwwroot` and `repo` being separate anymore, right? It's just putting files in `repo`?

2) `zotero.gpg` vs. `zotero-archive-keyring.gpg` — any reason to do the latter? I see that Ubuntu has an `ubuntu-archive-keyring.gpg`.

**Comment by dstillman** [comment]
3. Is it expected that I had to run `apt install ca-certificates` to get it to accept the repo? Without that it was saying the certificate for www.zotero.org is expired, which it obviously isn't.

**Comment by dstillman** [comment]
4. I think it makes more sense to use separate repos for release and beta — if you do `apt search zotero` you really shouldn't see `zotero-beta`. Can I just make `repo` the channel?

**Comment by dstillman** [comment]
5. What version of Debian/Ubuntu have you been building on?

I used `FROM ubuntu:22.04` for the Dockerfile, and then when trying to install the package on Ubuntu 20.04.2 LTS I got this:

```
[ubuntu:~] dan$  sudo apt install zotero-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 zotero-dev : Depends: libgdk-pixbuf-2.0-0 but it is not installable
E: Unable to correct problems, you have held broken packages.
```

This is the output of `apt-cache depends firefox-esr` in the Docker container:

```
firefox-esr
  Depends: lsb-release
  Depends: libatk1.0-0
  Depends: libc6
  Depends: libcairo-gobject2
  Depends: libcairo2
  Depends: libdbus-1-3
  Depends: libdbus-glib-1-2
  Depends: libfontconfig1
  Depends: libfreetype6
  Depends: libgcc-s1
  Depends: libgdk-pixbuf-2.0-0
  Depends: libglib2.0-0
  Depends: libgtk-3-0
  Depends: libharfbuzz0b
  Depends: libpango-1.0-0
  Depends: libpangocairo-1.0-0
  Depends: libpangoft2-1.0-0
  Depends: libstdc++6
  Depends: libx11-6
  Depends: libx11-xcb1
  Depends: libxcb-shm0
  Depends: libxcb1
  Depends: libxcomposite1
  Depends: libxcursor1
  Depends: libxdamage1
  Depends: libxext6
  Depends: libxfixes3
  Depends: libxi6
  Depends: libxrender1
  Depends: libxt6
  Recommends: xul-ext-ubufox
  Recommends: libcanberra0
  Recommends: libdbusmenu-glib4
  Recommends: libdbusmenu-gtk3-4
  Suggests: fonts-lyx
```

I'm trying again with `FROM ubuntu:20.04`, but isn't this sort of a problem in general? The current tarball is based on Firefox 60 ESR, but this is using the dependencies of firefox-esr 91.6.0esr+build1-0ubuntu0.22.04.1. How would that ever work?

Seems like the best we can do is hard-code a few things that will hopefully be available on most systems, no? (And then work on getting on a newer Firefox as soon as Z6 is out…)

**Comment by dstillman** [comment]
Oh, but after building on 20.04, it works. And the package name for libgdk-pixbuf is slightly different:

20.04:

```
Depends: gnupg, libatk1.0-0, libc6, libcairo-gobject2, libcairo2, libdbus-1-3, libdbus-glib-1-2, libfontconfig1, libfreetype6, libgdk-pixbuf2.0-0, libglib2.0-0, libgtk-3-0, libharfbuzz0b, libpango-1.0-0, libpangocairo-1.0-0, libpangoft2-1.0-0, libstdc++6, libx11-6, libx11-xcb1, libxcb-shm0, libxcb1, libxcomposite1, libxcursor1, libxdamage1, libxext6, libxfixes3, libxi6, libxrender1, libxt6
```

22.04:

```
Depends: gnupg, libatk1.0-0, libc6, libcairo-gobject2, libcairo2, libdbus-1-3, libdbus-glib-1-2, libfontconfig1, libfreetype6, libgdk-pixbuf-2.0-0, libglib2.0-0, libgtk-3-0, libharfbuzz0b, libpango-1.0-0, libpangocairo-1.0-0, libpangoft2-1.0-0, libstdc++6, libx11-6, libx11-xcb1, libxcb-shm0, libxcb1, libxcomposite1, libxcursor1, libxdamage1, libxext6, libxfixes3, libxi6, libxrender1, libxt6
```

Seems like they fixed the package name? Hopefully in a way where depending on libgdk-pixbuf2.0-0 still works in 22.04?

**Comment by retorquere** [comment]
> Can you explain rebuild.py?
> 
> Do I need that, or can I just hard-code `url` and `codename` in install.sh and put it on a server? And then run build.py and an upload script as part of the build process?

Yes. `rebuild.py` is irrelevant to zotero hosting, it's just something I used to prep the environment for `build.py` in the github action. The contents used to sit in one script, I split it into two specifically to lift out the zotero-hosting-relevant bits.

>     1. There's no point to `wwwroot` and `repo` being separate anymore, right? It's just putting files in `repo`?

That depends on how the build environment is laid out, and what you want to have show up as the channel. `repo` describes the build environment, `wwwroot` describes the hosting environment, the difference between those two shows up (a.o.) in the `Packages` file and as the channel in the `zotero.list`. If the two are the same you don't need them separately, but if they're the same, you'll get the `./` in the `zotero.list`, and you indicated you don't want that (for good reason I think). If you don't want `./` in the `zotero.list`, you need these two paths (or something similar).
 
>     2. `zotero.gpg` vs. `zotero-archive-keyring.gpg` — any reason to do the latter rather than the former? I see that Ubuntu has an `ubuntu-archive-keyring.gpg`.

I saw a recommendation for the latter option somewhere in de debian discussion list. I don't recall where, sorry. There is no material difference between these two.

> 3. Is it expected that I had to run `apt install ca-certificates` to get it to accept the repo? Without that it was saying the certificate for [www.zotero.org](http://www.zotero.org) is expired, which it obviously isn't.

It is unexpected to me, and since I run this script every two hours in a github action, it doesn't seem like it's something needed structurally. I'd venture to guess that this has to do with the distro picked for the docker image.

> 4. I think it makes more sense to use separate repos for release and beta — if you do `apt search zotero` you really shouldn't see `zotero-beta`. Can I just make `repo` the channel?

I'm not sure why you shouldn't see that? If I issue `apt search zo` I get a heap of hits across repos. But yes, if that is your preference, you can make `repo` the channel by a combination of `wwwroot`, `repo`, and which directory you pass to `build.py`. I would prefer to do some testing on my end to make sure it works (and give directions on how).

> 5. What version of Debian/Ubuntu have you been building on?

Ubuntu 20.04.3 LTS (github actions) and Ubuntu 21.10 (my own system)
 
> I used `FROM ubuntu:22.04` for the Dockerfile, and then when trying to install the package on Ubuntu 20.04.2 LTS I got 

22.04 is not out yet though

> [ubuntu:~] dan$  sudo apt install zotero-dev

this is the beta?

> The following packages have unmet dependencies:
>  zotero-dev : Depends: libgdk-pixbuf-2.0-0 but it is not installable
> E: Unable to correct problems, you have held broken packages.
> ```
> 
> This is the output of `apt-cache depends firefox-esr` in the Docker container:

This is indeed picked up in build.py, but it's strange that firefox-esr lists it as a dependency when it is not available in that environment.

> I'm trying again with `FROM ubuntu:20.04`, but isn't this sort of a problem in general? The current tarball is based on Firefox 60 ESR, but this is using the dependencies of firefox-esr 91.6.0esr+build1-0ubuntu0.22.04.1. How would that ever work?

It will work as long as the build environment has the same ESR as zotero uses. But I can disable this automatic pickup and add the required packages to the (existing) list of explicit packages.

> Seems like the best we can do is hard-code a few things that will hopefully be available on most systems, no? (And then work on getting on a newer Firefox as soon as Z6 is out…)

That will work.

> Seems like they fixed the package name? Hopefully in a way where depending on libgdk-pixbuf2.0-0 still works in 22.04?

They sometimes include compatibility packages, but I can't say whether that will happen here. If there is no compat package, I can deal with it in a `postinst`, which is sort of unfortunate, but it will work. Is firefox-esr installable in 22.04?

**Comment by retorquere** [comment]
How did you run this BTW? I have Docker installed but how to use it is still pretty fuzzy to me. If I can replicate the run steps it'd be easier for me to change `build.py` to accomodate what Zotero needs; it will also make it possible to test for the differences in 22.04.

**Comment by retorquere** [comment]
WRT to the differences between 20.10/20.04 and 22.04, another option than using a `postinst` is to just assume that the required libraries are already installed on the system, or just to assume that libgdk-pixbuf already is installed . This obviously works for the official tarball. I just preferred the dependencies to be explicit.

**Comment by dstillman** [comment]
Thanks for the responses. I'll let you know if I have other questions, but you don't need to do anything else here — I've already made lots of changes for our purposes, and just need to do some more testing.

**Comment by retorquere** [comment]
In light of [this](https://forums.zotero.org/discussion/comment/400259/#Comment_400259), a `postinst` could detect that the package is being installed on ChromeOS and could perhaps set `connector.url` to http://127.0.0.1:8080/.

**Comment by dstillman** [comment]
(Haven't forgotten about this. Couldn't get to it for Zotero 6, but should be soon.)

**Comment by retorquere** [comment]
I have in the meantime gotten the backblaze hosting up and running; I'm deprecating the GH-release and the SF mirror (although I will keep them up), but for the moment, I'm good.

Is there interest in the postinst detecting/adjusting for ChromeOS? Or is that best left to the user?

**Comment by eebrown** [comment]
Sorry if this is only tangentially related now, but I just ran deb2npm to see which node dependencies are already packaged and which are missing from Debian, which is relevant for any efforts/interest in packaging Zotero within Debian. Results below:

```npm2deb depends -r https://github.com/zotero/zotero/raw/master/package.json
Dependencies:
NPM                                               Debian
zotero (5.0.0)                                    None
├─ @citeproc-rs/wasm (^0.2.0)                     None
├─ ace-builds (^1.4.12)                           None
├─ bluebird (^3.5.1)                              node-bluebird (3.7.2+dfsg1+~3.5.36-1)
├─ classnames (^2.2.6)                            node-classnames (2.3.1-3)
├─ monaco-editor (^0.15.6)                        None
├─ prop-types (^15.8.0)                           node-prop-types (15.7.2+~15.7.4-2)
├─ react (^17.0.2)                                node-react (17.0.2+dfsg+~cs106.66.62-1)
├─ react-autosuggest (^10.1.0)                    None
│  ├─ es6-promise (^4.2.8)                        node-es6-promise (4.2.8-10)
│  ├─ prop-types (^15.7.2)                        node-prop-types (15.7.2+~15.7.4-2)
│  ├─ react-themeable (^1.1.0)                    None
│  │  └─ object-assign (^3.0.0)                   node-object-assign (4.1.1-6)
│  ├─ section-iterator (^2.0.0)                   None
│  └─ shallow-equal (^1.2.1)                      None
├─ react-dom (^17.0.2)                            None
│  ├─ loose-envify (^1.1.0)                       node-loose-envify (1.4.0+dfsg1-2)
│  ├─ object-assign (^4.1.1)                      node-object-assign (4.1.1-6)
│  └─ scheduler (^0.20.2)                         None
│     ├─ loose-envify (^1.1.0)                    node-loose-envify (1.4.0+dfsg1-2)
│     └─ object-assign (^4.1.1)                   node-object-assign (4.1.1-6)
├─ react-intl (^3.4.0)                            None
│  ├─ @formatjs/ecma402-abstract (1.11.3)         None
│  │  ├─ @formatjs/intl-localematcher (0.2.24)    None
│  │  │  └─ tslib (^2.1.0)                        node-tslib (2.3.1-2)
│  │  └─ tslib (^2.1.0)                           node-tslib (2.3.1-2)
│  ├─ @formatjs/icu-messageformat-parser (2.0.18) None
│  │  ├─ @formatjs/ecma402-abstract (1.11.3)      None
│  │  ├─ @formatjs/icu-skeleton-parser (1.3.5)    None
│  │  │  ├─ @formatjs/ecma402-abstract (1.11.3)   None
│  │  │  └─ tslib (^2.1.0)                        node-tslib (2.3.1-2)
│  │  └─ tslib (^2.1.0)                           node-tslib (2.3.1-2)
│  ├─ @formatjs/intl (2.1.0)                      None
│  │  ├─ @formatjs/ecma402-abstract (1.11.3)      None
│  │  ├─ @formatjs/fast-memoize (1.2.1)           None
│  │  │  └─ tslib (^2.1.0)                        node-tslib (2.3.1-2)
│  │  ├─ @formatjs/icu-messageformat-parser (2.0.18)None
│  │  ├─ @formatjs/intl-displaynames (5.4.2)      None
│  │  │  ├─ @formatjs/ecma402-abstract (1.11.3)   None
│  │  │  ├─ @formatjs/intl-localematcher (0.2.24) None
│  │  │  └─ tslib (^2.1.0)                        node-tslib (2.3.1-2)
│  │  ├─ @formatjs/intl-listformat (6.5.2)        None
│  │  │  ├─ @formatjs/ecma402-abstract (1.11.3)   None
│  │  │  ├─ @formatjs/intl-localematcher (0.2.24) None
│  │  │  └─ tslib (^2.1.0)                        node-tslib (2.3.1-2)
│  │  ├─ intl-messageformat (9.11.4)              None
│  │  │  ├─ @formatjs/ecma402-abstract (1.11.3)   None
│  │  │  ├─ @formatjs/fast-memoize (1.2.1)        None
│  │  │  ├─ @formatjs/icu-messageformat-parser (2.0.18)None
│  │  │  └─ tslib (^2.1.0)                        node-tslib (2.3.1-2)
│  │  └─ tslib (^2.1.0)                           node-tslib (2.3.1-2)
│  ├─ @formatjs/intl-displaynames (5.4.2)         None
│  ├─ @formatjs/intl-listformat (6.5.2)           None
│  ├─ @types/hoist-non-react-statics (^3.3.1)     None
│  │  ├─ @types/react (*)                         None
│  │  │  ├─ @types/prop-types (*)                 None
│  │  │  ├─ @types/scheduler (*)                  None
│  │  │  └─ csstype (^3.0.2)                      node-csstype (3.0.11-1)
│  │  └─ hoist-non-react-statics (^3.3.0)         None
│  │     └─ react-is (^16.7.0)                    None
│  ├─ @types/react (16 || 17)                     None
│  ├─ hoist-non-react-statics (^3.3.2)            None
│  ├─ intl-messageformat (9.11.4)                 None
│  └─ tslib (^2.1.0)                              node-tslib (2.3.1-2)
├─ react-select (^3.0.8)                          None
│  ├─ @babel/runtime (^7.12.0)                    node-babel (6.26.0+repack-3~bpo10+1)
│  ├─ @emotion/cache (^11.4.0)                    None
│  │  ├─ @emotion/memoize (^0.7.4)                None
│  │  ├─ @emotion/sheet (^1.1.0)                  None
│  │  ├─ @emotion/utils (^1.0.0)                  None
│  │  ├─ @emotion/weak-memoize (^0.2.5)           None
│  │  └─ stylis (4.0.13)                          None
│  ├─ @emotion/react (^11.1.1)                    None
│  │  ├─ @babel/runtime (^7.13.10)                node-babel (6.26.0+repack-3~bpo10+1)
│  │  ├─ @emotion/babel-plugin (^11.7.1)          None
│  │  │  ├─ @babel/helper-module-imports (^7.12.13)None
│  │  │  │  └─ @babel/types (^7.16.7)             node-babel (6.26.0+repack-3~bpo10+1)
│  │  │  ├─ @babel/plugin-syntax-jsx (^7.12.13)   node-babel (6.26.0+repack-3~bpo10+1)
│  │  │  ├─ @babel/runtime (^7.13.10)             node-babel (6.26.0+repack-3~bpo10+1)
│  │  │  ├─ @emotion/hash (^0.8.0)                None
│  │  │  ├─ @emotion/memoize (^0.7.5)             None
│  │  │  ├─ @emotion/serialize (^1.0.2)           None
│  │  │  │  ├─ @emotion/hash (^0.8.0)             None
│  │  │  │  ├─ @emotion/memoize (^0.7.4)          None
│  │  │  │  ├─ @emotion/unitless (^0.7.5)         None
│  │  │  │  ├─ @emotion/utils (^1.0.0)            None
│  │  │  │  └─ csstype (^3.0.2)                   node-csstype (3.0.11-1)
│  │  │  ├─ babel-plugin-macros (^2.6.1)          None
│  │  │  │  ├─ @babel/runtime (^7.12.5)           node-babel (6.26.0+repack-3~bpo10+1)
│  │  │  │  ├─ cosmiconfig (^7.0.0)               node-cosmiconfig (7.0.1+ds1-1)
│  │  │  │  └─ resolve (^1.19.0)                  node-resolve (1.20.0+~cs5.27.9-1)
│  │  │  ├─ convert-source-map (^1.5.0)           node-convert-source-map (1.8.0+~1.5.2-1)
│  │  │  ├─ escape-string-regexp (^4.0.0)         node-escape-string-regexp (4.0.0-2)
│  │  │  ├─ find-root (^1.1.0)                    None
│  │  │  ├─ source-map (^0.5.7)                   node-source-map (0.7.0++dfsg2+really.0.6.1-9)
│  │  │  └─ stylis (4.0.13)                       None
│  │  ├─ @emotion/cache (^11.7.1)                 None
│  │  ├─ @emotion/serialize (^1.0.2)              None
│  │  ├─ @emotion/utils (^1.1.0)                  None
│  │  ├─ @emotion/weak-memoize (^0.2.5)           None
│  │  └─ hoist-non-react-statics (^3.3.1)         None
│  ├─ @types/react-transition-group (^4.4.0)      None
│  │  └─ @types/react (*)                         None
│  ├─ memoize-one (^5.0.0)                        None
│  ├─ prop-types (^15.6.0)                        node-prop-types (15.7.2+~15.7.4-2)
│  └─ react-transition-group (^4.3.0)             node-react-transition-group (4.4.2+~cs4.4.4-1)
└─ url (^0.11.0)                                  node-url (0.11.0-5)

Build dependencies:
NPM                                               Debian
@babel/cli (^7.16.0)                              node-babel (6.26.0+repack-3~bpo10+1)
@babel/core (^7.16.5)                             node-babel (6.26.0+repack-3~bpo10+1)
@babel/eslint-parser (^7.16.5)                    None
@babel/eslint-plugin (^7.16.5)                    None
@babel/plugin-proposal-class-properties (^7.16.5) None
@babel/plugin-proposal-object-rest-spread (^7.16.5)None
@babel/plugin-proposal-optional-chaining (^7.16.5)None
@babel/plugin-proposal-unicode-property-regex (^7.16.5)None
@babel/preset-react (^7.16.5)                     node-babel (6.26.0+repack-3~bpo10+1)
@zotero/eslint-config (^1.0.6)                    None
babel-plugin-transform-es2015-modules-commonjs (^6.26.2)None
browserify (^14.5.0)                              node-browserify (17.0.0+ds-3)
chai (^4.1.2)                                     node-chai (4.3.6~ds1+~cs9.7.12-3)
chai-as-promised (^7.1.1)                         node-chai-as-promised (7.1.1+~cs7.1.4-2)
chokidar (^2.1.2)                                 node-chokidar (3.5.3-1)
co-mocha (^1.2.1)                                 None
colors (^1.4.0)                                   colors.js (1.4.0-3)
eslint (^8.5.0)                                   None
eslint-plugin-react (^7.28.0)                     None
fs-extra (^3.0.1)                                 node-fs-extra (10.0.1+~9.0.13-2)
globby (^6.1.0)                                   node-globby (12.0.2+~cs16.24.35-3)
jspath (^0.4.0)                                   None
mocha (^3.5.3)                                    node-mocha (9.2.2+ds1+~cs28.3.8-1)
multimatch (^2.1.0)                               node-multimatch (5.0.0-2)
react-virtualized (^9.22.3)                       None
sass (^1.28.0)                                    None
sinon (^7.3.2)                                    node-sinon (13.0.1+ds+~cs71.22.21-2)
universalify (^0.1.1)                             node-universalify (2.0.0-3)
```
